### PR TITLE
refactor: move partition related code to partition manager

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
Move partition rule related code to partition manager.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#901

### CODE REVIEW

File: src/partition/src/range.rs [hunk 1]
```diff
@@ -82,16 +83,16 @@ impl RangePartitionRule {
...
     }
 
-    pub(crate) fn all_regions(&self) -> &Vec<RegionNumber> {
+    pub fn all_regions(&self) -> &Vec<RegionNumber> {
         &self.regions
     }
 
-    #[cfg(test)]
-    pub(crate) fn bounds(&self) -> &Vec<Value> {
+    // #[cfg(test)]
```
Comment (MichaelScofield) [inline src/partition/src/range.rs:94]:
```suggestion
```

**Review by MichaelScofield — APPROVED** [review]

File: src/partition/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+[dependencies]
+common-error = { path = "../common/error" }
+common-query = { path = "../common/query" }
+datafusion.workspace = true
+datafusion-common.workspace = true
+datafusion-expr.workspace = true
+datatypes = { path = "../datatypes" }
+meta-client = { path = "../meta-client" }
+moka = { version = "0.9", features = ["future"] }
+snafu = { version = "0.7", features = ["backtraces"] }
```
Comment (evenyag) [inline src/partition/Cargo.toml:18]:
```suggestion
snafu.workspace = true
```

File: src/partition/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+store-api = { path = "../store-api" }
+serde = "1.0"
```
Comment (evenyag) [inline src/partition/Cargo.toml:20]:
```suggestion
serde.workspace = true
```

File: src/partition/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![feature(assert_matches)]
+
+use std::sync::Arc;
+
+use crate::partition::PartitionRule;
```
Comment (evenyag) [inline src/partition/src/lib.rs:19]: Also pub use this?

File: src/partition/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+pub mod columns;
+pub mod error;
+pub mod manager;
+pub mod partition;
+pub mod range;
+pub mod route;
+pub mod splitter;
+
+pub type PartitionRuleRef = Arc<dyn PartitionRule<Error = error::Error>>;
```
Comment (evenyag) [inline src/partition/src/lib.rs:29]: Is it still necessary to use associated type for this error?

File: src/partition/src/manager.rs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+pub type PartitionRuleManagerRef = Arc<PartitionRuleManager>;
+
+pub struct PartitionRuleManager {
+    table_routes: Arc<TableRoutes>,
+}
+
+/// Provides methods to find regions by:
+/// - values (in case of insertion)
+/// - filters (in case of select, deletion and update)
+impl PartitionRuleManager {
```
Comment (evenyag) [inline src/partition/src/manager.rs:43]: What about moving them to the `struct` definition?
```rust
/// Provides methods to find regions ......
pub struct PartitionRuleManager {}
```

**Pull request merged**